### DISCUSSION TITLE
Refactoring Primitive Obsession

### DISCUSSION THREAD

**Discussion by gregorriegler** [openrewrite/rewrite#689] [Q&A]
Hi,

I am trying to get started writing Recipes and struggling.
What I would like to do is detect a primitive collection such as `List<Product>`
and turn it into a `ProductList` that encapsulates the list via composition.

For that I see the following steps necessary:

1. Detect and remember all methods of `List<Product>` that were actually used such as `add`, and `remove`.
2. Create a new domain-specific Class `ProductList` that holds `List<Product>` as a field and delegates methods that were previously used.
3. Change all Type occurrences of `List<Product>` to `ProductList`

For 3. I could use the preexisting `ChangeType` recipe, however, then I needed to make a composite of many Recipes. This is already my first Question. Can I make a composite of many Recipes and test it as well as use it as if it was just a single one?
**update: just noticed that recipes implement a chain of responsibility.**

How would you go about solving this problem overall?
How could I find all used methods of List<Product> to use them in a new JavaTemplate that created the WrapperClass? Ideas welcome.

greetings,
Gregor

**Comment by taravirginillo** [comment]
Hey Greg! I'm dealing with a similar issue right now. Did you figure out a composite recipe that would work for you? Still trying to figure out a recipe to detect methods of a given primitive. I also was considering using `ChangeType` along with a composite of other recipes. Did that method succeed for you? Let me know if there's been any updates since you've posted this!

**Reply by gregorriegler** [reply]
Hi Tara, I am still working on this with a friend of mine. We started way simpler than the use case I described in my first post.
Made some successes, but it's much more complicated than I anticipated. Love to share though. Forum is not the best medium for this I think, its not even finished.

**Reply by timtebeek** [reply]
Hi @gregorriegler ! Sorry to see you did not get a reaction sooner; I've now hooked up GitHub discussions to our Slack so we get notified about any new discussions.

Interesting use case; curious to see how far you've some since and how we can best support you going further. Welcome to [drop by our Slack](https://join.slack.com/t/rewriteoss/shared_invite/zt-nj42n3ea-b~62rIHzb3Vo0E1APKCXEA); the team is quite active there to help you further.

**Comment by timtebeek** [comment]
Closing as stale; feel free to open a new discussion if this is still relevant.
